### ISSUE TITLE
Median of Medians Algorithm

### ISSUE THREAD

**Issue by Nils-Goldmann** [TheAlgorithms/Java#2830]
Hello,

I would like to add the median of medians algorithm (see e.g. https://en.wikipedia.org/wiki/Median_of_medians). So far, I have not seen any implementation in Java, so this could be a nice addition.

Example test case: 

    medianOfMedians([2,87,18,20,15,5,66,0,43,37,12]) --> 18

**Comment by Nils-Goldmann** [comment]
A colleague and I have come up with a rough plan how to proceed: We will write a program containing the algorithm as well as a test suite. For the tests, we will probably resort to a main function with a series of asserts, since this is most consistent with the project's current structure. We plan on putting the program into the maths directory, as there already is a median program. Also, we will add a link to DIRECTORY.md.

If we have the time, we may also add a description/explanation for the algorithm as well as a running time and space complexity analysis.

Please let us know if you have any advice, additional requests etc :)

**Comment by siriak** [comment]
Do you know how unit tests can be added? I'd love to include them instead of main() and run them in the CI pipeline somehow

**Comment by Nils-Goldmann** [comment]
I have added a PR on this issue: https://github.com/TheAlgorithms/Java/pull/2860. It would be great if you had time to review it :)

We renamed the feature after the actual algorithm shown at https://en.wikipedia.org/wiki/Median_of_medians, namely QuickSelect. Median of Medians is an independent algorithm per se, however QuickSelect is one of the most common applications. It appeared most sensible to us to use the same algorithm as in the reference. Hence, we renamed the feature accordingly and created a new branch for it.
